### ISSUE TITLE
Out of place NameOf suggestion

### ISSUE THREAD

**Issue by mc0re** [code-cracker/code-cracker#460]
The following line:

<CodeAnalysis.SuppressMessage("Style", "CC0013:Use Ternary operator.", Justification:="...")>

produces message CC0021 for word "Style", suggesting to use "NameOf(Style)" instead. This is certainly not a place where you might want to rename anything.

**Comment by adraut** [comment]
@mc0re Your sample text did not get saved.  Could try posting it again?

**Comment by giggio** [comment]
@adraut He had the code without scaping, I updated it.

**Comment by giggio** [comment]
@mc0re NameOfAnalyzer will check for anything in context. If an identifier `Style` is context, it will offer to use nameof there. Why is that a problem?

**Comment by mc0re** [comment]
Because exactly in this case, `Style` is a string, a keyword, and not a name of the `Style` class. And it can't be a name for `SupressMessageAttribute`. May be, one could mark such places somehow, so the NameOf check skips them?

**Comment by giggio** [comment]
Style is not a keyword. It might be coming from a namespace. The discussion would be what types of identifiers we would like to ignore, and in which situations. Do we want to ignore namespaces, for example? And why?

**Comment by mc0re** [comment]
I meant, `Style` there is a fixed string (like a keyword in some language), independent of the class or other names.
I suggest some markup, e.g. an attribute `NoNameOfAttribute` or some comment, which defines, that for a certain method/constructor parameter or for a certain field/property of a class NameOf checks should be disabled.

**Comment by giggio** [comment]
To be able to use an attribute we would need to create an assembly that people would then reference on their projects. I am not sure that we want to do that, but I am open to work with the idea.
On C# you can disable the analyzer on that line easily, like that:

````csharp
#pragma warning disable CC0021
[CodeAnalysis.SuppressMessage("Style", "CC0013:Use Ternary operator.", Justification:="...")]
#pragma warning restore CC0021
````

But that does not exist on VB.

**Comment by mc0re** [comment]
What about a comment? I know, Roslyn parser can also traverse comments. So something specially formatted could be spotted.

**Comment by giggio** [comment]
Hey, you are right. We could do that. We could look into it like that:

````vb
'disable CC0021
<CodeAnalysis.SuppressMessage("Style", "CC0013:Use Ternary operator.", Justification:="...")>
````

That is actually a great idea. We could maybe implement that for all VB analyzers, always check for this comment before. C# has `pragma warning disable`, VB could have that, at least for CodeCracker.
I like it.

**Comment by mc0re** [comment]
Well, that's great, but not exactly what I was thinking about :-)
I was thinking more like disabling a warning altogether, like:

```vb
Public Class A
    ' CC0021 disable always for Keyword
    Public Property Keyword As String
```

and maybe:
`' CC0021 disable always for CodeAnalysis.SuppressMessage.Argument1`

**Comment by giggio** [comment]
It would be easy to implement it for a comment on a direct method, all we have to do is check the trivia.
But for anything else, it would be **really** hard. If we had a loose comment on a .vb file, to be able to get it to work, we would have to gather **all** comments from the whole compilation, and only on compilation end we could then decide what analyzers should report or not. This would actually touch on another bug, detailed on https://github.com/code-cracker/code-cracker/pull/292#issuecomment-77065161.

**Comment by mc0re** [comment]
One could say that such comments should be defined either in direct relation to the attributed entity (e.g. right before the method in question) or in a predefined file (e.g. "CCAnnotations.vb").

**Comment by giggio** [comment]
Even in a predefined file we would have problems, because the compilation might not had run yet. It would cause the same problem I reported on the comment just before.

**Comment by mc0re** [comment]
You're right. The only way to assure the compilation order is to introduce a dependency from this predefined file, e.g. define something in it to be used in every file, which needs the comments. It's probably too much to ask the programmers to do...

**Comment by mirhagk** [comment]
I'm not sure if this issue is resolved or not but I noticed that code cracker does not verify on interpolated strings (`$"Style"`) so perhaps this could be a solution? Perhaps it's worth considering whether it should also ignore literal strings (`@"Style"`) which would give this a pretty clean solution.

Tbh I'm not 100% confident on this, just thought I'd mention that C# has 3 string types and perhaps code cracker could use the different types to decide whether it's a candidate or not

**Comment by giggio** [comment]
@mirhagk I don't think we should be differentiating string types, as this is not really clear for the user.
Interpolated string are not easy to check, only those with no real interpolations could be replaced by nameof, and those do not make sense to even exist.

**Comment by mirhagk** [comment]
Yeah I just mentioned because it was possible to use as a workaround to get rid of false positive situations.

**Comment by RachelDavids** [comment]
![image](https://cloud.githubusercontent.com/assets/18124189/14914322/f1a12ce0-0e00-11e6-8875-6318cff2271e.png)

An example with more false positives in VB. These are simply English names and do not necessarily have direct relationship with any entity in the code. Whilst it is possible they match the properties of the object being displayed, the applied fix does not recognize this (how can it - the population occurs elsewhere...) and makes the code dependent on an entity it has no relationship with.

**Comment by paul1956** [comment]
I modified the CodeFix to ignore anything in Microsoft libraries and provided a way to list other Namespaces to ignore. I started with a simple Ignore list that I load in the VSIX then I added a menu to update the list. So far there too many usability issues to release it. Just ignoring three Namespaces eliminates almost all the noise.
In VSIX
```
       Private _csvIgnoreList As String = "Microsoft,System,MS"
        Public Property csvIgnoreList As String Implements INameOfSettings.csvIgnoreList
            Get
                Return _csvIgnoreList
            End Get
            Set(value As String)
                _csvIgnoreList = value
            End Set
        End Property
```
In CodeAnalyzer
Then I modified GetProgramElementNameThatMatchesStringLiteral to check for things on the list. There is some code read and translate the CSV string to String array not shown.
```
                If symbol.ToDisplayString.StartsWithAny(_NameSpaceIgnoreList) Then
                    Return Nothing
                End If
                If symbol.ContainingAssembly?.Name?.StartsWithAny(_NameSpaceIgnoreList) Then
                    Return Nothing
                End If
```
Eventually I want to use JSON to handle this and do some caching to avoid reading the files repeatedly but still detect external changes.

**Comment by mc0re** [comment]
Is it released?
